fpm/src/fpm.f90

```diff
@@ -166,25 +172,206 @@ subroutine build_model(model, settings, package, error)
-            & '-J'//join_path(model%output_directory,model%package_name)
-    else
-        model%output_directory = 'build/gfortran_debug'
-        model%fortran_compile_flags = ' -Wall -Wextra -Wimplicit-interface  -fPIC -fmax-errors=1 -g '// &
-                                      '-fbounds-check -fcheck-array-temporaries -fbacktrace '// &
-                                      '-J'//join_path(model%output_directory,model%package_name)
-    endif
+   model%fortran_compiler=settings%compiler
+
+   model%output_directory = join_path('build',model%fortran_compiler//'_'//settings%build_name)
+
+   if(settings%compiler.eq.'')then
+       model%fortran_compiler = 'gfortran'
+   else
+       model%fortran_compiler = settings%compiler
+   endif
+
+! #TODO: Choose flags and output directory based on cli settings & manifest inputs
+! special reserved names "debug" and "release" are for supported compilers with no user-specified compile or load flags
+
+! vendor            Fortran   C         Module output   Module include OpenMP    Free for OSS
+!                   compiler  compiler  directory       directory
+! Gnu               gfortran   gcc     -J              -I            -fopenmp   X
+! Intel             ifort      icc     -module         -I            -qopenmp   X
+! Intel(Windows)    ifort      icc     /module:path    /I            /Qopenmp   X
+! Intel oneAPI      ifx        icx     -module         -I            -qopenmp   X
+! PGI               pgfortran  pgcc    -module         -I            -mp        X
+! NVIDIA            nvfortran  nvc     -module         -I            -mp        X
+! LLVM flang        flang      clang   -module         -I            -mp        X
+! LFortran          lfortran   ---     ?               ?             ?          X
+! Lahey/Futjitsu    lfc        ?       -M              -I            -openmp    ?
+! NAG               nagfor     ?       -mdir           -I            -openmp    x
+! Cray              crayftn    craycc  -J              -I            -homp      ?
+! IBM               xlf90      ?       -qmoddir        -I            -qsmp      X
+! Oracle/Sun        ?          ?       -moddir=        -M            -xopenmp   ?
+! Silverfrost FTN95 ftn95      ?       ?               /MOD_PATH     ?          ?
+! Elbrus            ?          lcc     -J              -I            -fopenmp   ?
+! Hewlett Packard   ?          ?       ?               ?             ?          discontinued
+! Watcom            ?          ?       ?               ?             ?          discontinued
+! PathScale         ?          ?       -module         -I            -mp        discontinued
+! G95               ?          ?       -fmod=          -I            -fopenmp   discontinued
+! Open64            ?          ?       -module         -I            -mp        discontinued
+! Unisys            ?          ?       ?               ?             ?          discontinued
+
+    select case(settings%build_name//'_'//settings%compiler)
```

Comment:
Can we move most of the compiler logic to a separate module (`src/fpm/compiler.f90`), this would keep the toplevel fpm module clean from this kind of logic and we have a dedicated place to expand on this logic.

Reply:
Was thinking same thing;  was thinking of defining a compiler type type instead of select; but as you mentioned just moving it to a seperate module simplifies that.

Reply:
So the file fpm_compiler.f90 was added with a routine to set default compiler flags that encapsulates that function for
further expansion.

The only change to the user interface is the addtion of the --compiler COMPILER_NAME switch and the use of the environment variable FPM_COMPILER to select the default compiler. The full value of COMPILER_NAME is used as the compiler command (allowing for full pathnames) but the basename of that pathname must match one of the pre-defined compiler names and that name is used to create the build directory name.  In general (for now)  the names are essentially
limited to those in fpm_compiler.f90

No mechanism for the user to set custom flags is included. 

Encoding the possible full pathname that might be in FPM_COMPILER in a way similiar to what h-fpm does for compiler flags could be a way to allow for a more generic mechanism for creating compiler-specific directories.

No use is made of the FC and FFLAGS variables or for allowing user-specific compiler options.

For now, it is assumed COMPILER is a Fortran Compiler.  There is the question of how best to allow for a list of encoders for a possible list of supported languages.

In fpm_compiler.f90 if user-defined type is created that defines a compilers options it would be simpler to be able to print a table of the types to be able to show users the acceptable names and their flags but currently there is no way to query what compilers are supported.  Either way if anyone has suggested flag values for the various compilers included let me know.

The verbose switch is available for any command, although I didn't add much logic to use it.  One question is whether the external commands should only echo if the --verbose switch is on; or conversely if there should be a --quiet switch (I see several other packages have both a  -v|-V an a -q switch to allow for quiet, normal, and verbose).


## experiences
As is, I find it very useful to build using multiple compilers. I find I tend to use the variable and aliases fpm-i, fpm-n, and fpm-g for three different compilers myself

Reply:
@urbanjost since you mentioned that you are testing several compilers with aliases. Maybe you know about response files supported by some command line tools, I opened an issue at M_CLI2 regarding this https://github.com/urbanjost/M_CLI2/issues/5. This could allow to store compiler flags in files rather than shell aliases.